File: solutions/2026-04-16/cpp/2026-04-16-closest-equal-element-queries.cpp
```cpp
// Problem: Closest Equal Element Queries
// Problem Link: https://leetcode.com/problems/closest-equal-element-queries/
// Approach:
// We need to efficiently find the closest equal element for each query in a circular array.
// For each number in `nums`, we can store all its indices in a sorted list.
// A hash map (unordered_map in C++) is suitable for this, where keys are the numbers and values are sorted vectors of their indices.
// After pre-processing, for each query `queries[i]`:
// 1. Get the value `val = nums[queries[i]]`.
// 2. Get the list of indices `indices` for `val` from the hash map.
// 3. If `indices` has only one element (which is `queries[i]` itself), then no other equal element exists, return -1.
// 4. Otherwise, we need to find the closest index in `indices` to `queries[i]`.
//    Since `indices` is sorted, we can use binary search (`lower_bound`) to find the position of `queries[i]` (or the first element greater than or equal to it).
//    Let `it` be the iterator returned by `lower_bound`.
//    - If `*it` is exactly `queries[i]`, then `it` points to the current element's index.
//      - We check the element before `it` (if `it` is not the beginning) and the element after `it` (if `it` is not the end) to find the minimum distance.
//      - The circular distance needs to be calculated: `min(abs(current_index - other_index), n - abs(current_index - other_index))`, where `n` is the length of `nums`.
//    - If `*it` is greater than `queries[i]`, then `it` points to the first element greater than `queries[i]`.
//      - We check the element before `it` (if `it` is not the beginning) and `*it` (if `it` is not the end) for the minimum distance.
// 5. The minimum of these circular distances is the answer for the current query.
// Time Complexity:
// - Pre-processing: O(N log N) in the worst case if all elements are the same and we need to sort the indices, but typically O(N) on average for hash map insertion and then O(N log K) where K is the maximum number of occurrences of any element if we sort each list. More precisely, if we just store them and they happen to be sorted, it's O(N). If we need to sort, it's Sum(Ni log Ni) where Ni is the count of each number. In the worst case, if all numbers are the same, this is N log N. If numbers are unique, this is N.
// - Querying: For each query, binary search on a list of indices takes O(log K) where K is the number of occurrences of that element. Since there are Q queries, the total query time is O(Q log K_max), where K_max is the maximum number of occurrences of any element.
// - Overall: O(N + Q log N) or O(N log N + Q log N) depending on whether sorting is needed for indices. Given constraints, N log N is dominant for pre-processing if needed. O(N + Q log N) is a more accurate general bound if indices are added in order.
// Space Complexity:
// - O(N) to store the indices of each number in the hash map.
//
class Solution {
public:
    std::vector<int> closestDistance(std::vector<int>& nums, std::vector<int>& queries) {
        int n = nums.size();
        // Map to store indices for each number. Key: number, Value: sorted list of indices.
        std::unordered_map<int, std::vector<int>> val_to_indices;

        // Populate the map with indices for each number.
        for (int i = 0; i < n; ++i) {
            val_to_indices[nums[i]].push_back(i);
        }

        std::vector<int> answer;
        answer.reserve(queries.size()); // Reserve space for efficiency

        for (int query_idx : queries) {
            int current_val = nums[query_idx];
            const auto& indices = val_to_indices[current_val];

            // If there's only one occurrence of this value, no other equal element exists.
            if (indices.size() == 1) {
                answer.push_back(-1);
                continue;
            }

            int min_dist = n; // Initialize with a value larger than any possible distance

            // Find the position of query_idx in the sorted list of indices for current_val.
            // lower_bound returns an iterator to the first element not less than query_idx.
            auto it = std::lower_bound(indices.begin(), indices.end(), query_idx);

            // Check the element at `it` (if `it` is not the end)
            if (it != indices.end()) {
                int other_idx = *it;
                // Calculate circular distance
                int dist = std::min(std::abs(query_idx - other_idx), n - std::abs(query_idx - other_idx));
                min_dist = std::min(min_dist, dist);
            }

            // Check the element before `it` (if `it` is not the beginning)
            if (it != indices.begin()) {
                // Move iterator back to the previous element
                --it;
                int other_idx = *it;
                // Calculate circular distance
                int dist = std::min(std::abs(query_idx - other_idx), n - std::abs(query_idx - other_idx));
                min_dist = std::min(min_dist, dist);
            }

            // If min_dist is still n, it means something went wrong or no valid index was found.
            // This case should be covered by the indices.size() == 1 check, but as a safeguard.
            if (min_dist == n) {
                 answer.push_back(-1);
            } else {
                 answer.push_back(min_dist);
            }
        }

        return answer;
    }
};
```

File: solutions/2026-04-16/java/2026-04-16-closest-equal-element-queries.java
```java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Collections;

class Solution {
    /**
     * Problem: Closest Equal Element Queries
     * Link: https://leetcode.com/problems/closest-equal-element-queries/
     *
     * Approach:
     * 1. Preprocessing:
     *    - Create a map where keys are the numbers in `nums` and values are sorted lists of indices where that number appears.
     *    - Iterate through `nums` and populate this map. For each number, add its index to the corresponding list.
     * 2. Query Processing:
     *    - For each query `queries[i]`:
     *      - Get the value `val = nums[queries[i]]` and the current index `currIndex = queries[i]`.
     *      - If the list of indices for `val` in the map has only one element (meaning `val` appears only at `currIndex`), the answer is -1.
     *      - Otherwise, find the indices in the sorted list that are closest to `currIndex`.
     *      - Use binary search (`Collections.binarySearch`) to find `currIndex` in the list of indices for `val`.
     *      - If `currIndex` is found, its insertion point will be negative. The elements closest to it will be at `insertionPoint - 1` and `insertionPoint` (if they exist within bounds).
     *      - If `currIndex` is not found, `binarySearch` returns `-(insertionPoint) - 1`. The elements closest to it will be at `insertionPoint - 1` and `insertionPoint` (if they exist within bounds).
     *      - Calculate the circular distance to the closest indices. The circular distance between two indices `a` and `b` in an array of length `n` is `min(abs(a - b), n - abs(a - b))`.
     *      - The minimum of these circular distances is the answer for the query.
     *
     * Time Complexity:
     * - Preprocessing: O(N log N) if we sort the indices for each number. However, if we just append and then sort only if needed, or if we process queries in a specific order, it can be improved. A simpler approach is to populate the map with indices and then sort each list of indices. This takes O(N) to populate and then O(N log N) in the worst case if all elements are the same and we sort all lists. A better way: O(N) to build the map.
     * - Query Processing: For each query, we perform a binary search on a list of indices. In the worst case, a number can appear O(N) times, making the binary search O(log N). So, for Q queries, it's O(Q log N).
     * - Overall: O(N + Q log N). If sorting each list of indices is done upfront, it becomes O(N log N + Q log N). The current implementation sorts as needed, but `Collections.binarySearch` implicitly handles sorted lists. The crucial part is that the `List<Integer>` is sorted for binary search.
     * - Let's refine the preprocessing:
     *   - Building the map: O(N)
     *   - Sorting all index lists: If a number appears `k` times, sorting its list is O(k log k). Summing over all unique numbers, the total sorting time is at most O(N log N) if all elements are the same, or O(N) if elements are mostly unique. A tighter bound is O(N log N) in the worst case if we sort each list.
     *   - Querying: O(Q log N)
     *   - Total: O(N log N + Q log N) after sorting all lists upfront.
     *
     * Space Complexity:
     * - O(N) to store the map of indices.
     */
    public int[] closestEquidistantQueries(int[] nums, int[] queries) {
        int n = nums.length;
        // Map to store indices for each number: number -> sorted list of indices
        Map<Integer, List<Integer>> valToIndexList = new HashMap<>();

        // Preprocessing: Populate the map with indices for each value
        for (int i = 0; i < n; i++) {
            valToIndexList.computeIfAbsent(nums[i], k -> new ArrayList<>()).add(i);
        }

        // Sort the index lists for each value to enable binary search
        for (List<Integer> indices : valToIndexList.values()) {
            Collections.sort(indices);
        }

        int[] answer = new int[queries.length];

        // Process each query
        for (int i = 0; i < queries.length; i++) {
            int currIndex = queries[i];
            int val = nums[currIndex];

            List<Integer> indices = valToIndexList.get(val);

            // If the value appears only once, there's no other equal element
            if (indices.size() == 1) {
                answer[i] = -1;
            } else {
                // Find the position of the current index in the sorted list of indices
                int searchResult = Collections.binarySearch(indices, currIndex);

                int insertionPoint;
                if (searchResult >= 0) {
                    // currIndex was found at searchResult
                    insertionPoint = searchResult;
                } else {
                    // currIndex was not found, binarySearch returns (-(insertion point) - 1)
                    // The insertion point is where it *would* be inserted to maintain order
                    insertionPoint = -(searchResult + 1);
                }

                int minDistance = Integer.MAX_VALUE;

                // Check the element immediately before the insertion point (if it exists)
                if (insertionPoint > 0) {
                    int prevIndex = indices.get(insertionPoint - 1);
                    // Calculate circular distance
                    int distance = Math.min(Math.abs(currIndex - prevIndex), n - Math.abs(currIndex - prevIndex));
                    minDistance = Math.min(minDistance, distance);
                }

                // Check the element immediately at or after the insertion point (if it exists)
                // If currIndex was found, insertionPoint is its index. We need to check the next element.
                // If currIndex was not found, insertionPoint is where it would be inserted. We check the element at insertionPoint.
                int nextIndexToCheck = (searchResult >= 0) ? insertionPoint + 1 : insertionPoint;
                if (nextIndexToCheck < indices.size()) {
                    int nextIndex = indices.get(nextIndexToCheck);
                    // Calculate circular distance
                    int distance = Math.min(Math.abs(currIndex - nextIndex), n - Math.abs(currIndex - nextIndex));
                    minDistance = Math.min(minDistance, distance);
                }
                
                // The problem statement implies that we are guaranteed to find at least one other index if size > 1.
                // If minDistance is still MAX_VALUE here, it means something is wrong or the only element is currIndex itself (handled by size == 1 check).
                answer[i] = minDistance;
            }
        }

        return answer;
    }
}
```